subsys: debug/coredump: add a primitive coredump mechanism

dcpleung opened the pull request:
This adds a very primitive coredump mechanism under subsys/debug
where during fatal error, register and memory content can be
dumped to coredump backend. One such backend utilizing log
module for output is included. Once the coredump log is converted
to a binary file, it can be used with the ELF output file as
inputs to an overly simplified implementation of a GDB server.
This GDB server can be attached via the target remote command of
GDB and will be serving register and memory content. This allows
using GDB to examine stack and memory where the fatal error
occurred.

dcpleung committed:
tests: add a test for coredump

This adds a simple test for coredump.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

dcpleung committed:
debug/coredump: add a primitive coredump mechanism

This adds a very primitive coredump mechanism under subsys/debug
where during fatal error, register and memory content can be
dumped to coredump backend. One such backend utilizing log
module for output is included. Once the coredump log is converted
to a binary file, it can be used with the ELF output file as
inputs to an overly simplified implementation of a GDB server.
This GDB server can be attached via the target remote command of
GDB and will be serving register and memory content. This allows
using GDB to examine stack and memory where the fatal error
occurred.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

dcpleung committed:
coredump: add support for x86 and x86_64

This adds the necessary bits to enable coredump for x86
and x86_64.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

dcpleung commented:
Still trying to figure out how to test this... as the registers and memory content are going to be different between runs.

andrewboie commented:
Neat! Curious what requires userspace to be disabled?

dcpleung commented:
> Neat! Curious what requires userspace to be disabled?

I haven't really tested with userspace so this is there to avoid enabling coredump, as I remember seeing an error before where supervisor thread was not allowed to read userspace data. This would cause a page fault when trying to dump memory.

andrewboie commented:
> I haven't really tested with userspace so this is there to avoid enabling coredump, as I remember seeing an error before where supervisor thread was not allowed to read userspace data. This would cause a page fault when trying to dump memory.

Supervisor mode should have read access to all system RAM. There should be no cases where user mode can read a page but the kernel can't. If that's not the case please file a bug.

dcpleung commented:
> Supervisor mode should have read access to all system RAM. There should be no cases where user mode can read a page but the kernel can't. If that's not the case please file a bug.

Tested with the sample userspace apps and I was able to dump memory. So this can be used with userspace now.

dcpleung committed:
sanitycheck: need to wait for harness to process QEMU outputs

There is a potential that QEMUHandler.handle() returns before
the testing harness finishes processing all the output from QEMU.
Simply wait for the harness thread to finish before returning.

Also, fix the return code in the debug message as it should be
the return code from Popen().

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

dcpleung committed:
sanitycheck: allow QEMU to crash without failing

Adds some code to allow QEMU to crash without failing the test.
This is required for testing coredump code as it will certainly
cause QEMU to crash.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

dcpleung committed:
coredump: add support for ARM Cortex-M

This adds the necessary bits in arch code, and Python scripts
to enable coredump support for ARM Cortex-M.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

ceolin commented on scripts/coredump/coredump_serial_log_parser.py:
```diff
@@ -0,0 +1,99 @@
+    go_parse_line = False
+    bytes_written = 0
+    for line in infile.readlines():
+        if line.find(COREDUMP_BEGIN_STR) >= 0:
```
shouldn't `go_parse_line` be initialized to `False` in the beginning of this loop ? Otherwise you may have a situation where the previous line was `COREDUMP_BEGIN_STR`, consequently `go_parse_line` is `True` now, and the current line does not trigger any of these `if` causing the line to be invalid parsed.

Is that a possible scenario ?

dcpleung replied:
This is mainly to skip all the lines before beginning of coredump log, since log is enclosed with `BEGIN`/`END` line pair. In other words, this ensures only lines are parsed after `BEGIN` has been seen (and until `END`).

ceolin commented on scripts/coredump/coredump_gdbserver.py:
```diff
@@ -0,0 +1,127 @@
+LOGGING_FORMAT = "[%(levelname)s][%(name)s] %(message)s"
+
+GDBSERVER_HOST = ""
+GDBSERVER_PORT = 1234
```
very very minor, but the port value could be an argument, that is the default port and one may already being debugging a qemu instance, for example.

andrewboie replied:
agreed we should make this an argument, with a default of 1234

dcpleung replied:
Made this as an argument.

ceolin commented on scripts/coredump/coredump_gdbserver.py:
```diff
@@ -0,0 +1,127 @@
+        conn.close()
+
+    gdbserver.close()
+
```
```
elff.close()
logf.close()
```
or they are automatically closed ? Never knows how it works on python.

dcpleung replied:
Normally all file descriptors are closed once a program terminates. I will add the close commands just in case.

ceolin commented on include/debug/coredump.h:
```diff
@@ -0,0 +1,139 @@
+ * @fn void z_coredump(unsigned int reason, const z_arch_esf_t *esf, struct k_thread *thread);
+ * @brief Perform coredump during z_fatal_error().
+ *
+ * This should be called with z_fatal_error() to generate coredump.
```
similar to a comment @andrewboie did in my gdbstub pr, shouldn't this API be in a private header ?

andrewboie replied:
yeah if this is really intended for `z_fatal_error()` it should be private.

but if you actually meant for it to be called from an implementation of `k_sys_fatal_error_handler()` (which is customizable) then it can remain and the comment should be updated

dcpleung replied:
The idea is that this can also be called on demand outside of fatal errors if one is determined to do so. GDB allows you to do core dump while debugging, so this is (somehow) mirroring that ability. Therefore it is under `include/` but has a prefix `z_`.

ceolin commented on doc/guides/debugging/coredump.rst:
```diff
@@ -0,0 +1,347 @@
+Here are the options to enable output backends for core dump:
+
+* ``DEBUG_COREDUMP_BACKEND_LOGGING``: use log module for core dump output.
+* ``DEBUG_COREDUMP_BACKEND_NULL``: fallback core dump backend if other
```
what is the purpose of having it ? Wouldn't just be better not enable coredump in this case ?

dcpleung replied:
I had it at first thinking I would write a build only test for this. But now this is not really needed, so removing it...
